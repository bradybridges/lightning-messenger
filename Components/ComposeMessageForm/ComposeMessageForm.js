import React, { Component } from 'react';
import { Text, TextInput, View, Image, TouchableOpacity, StyleSheet, Dimensions, AsyncStorage } from 'react-native';
import * as firebase from 'firebase';
import 'firebase/firestore';
const sendImg = require('../../assets/send.png');
import * as constants from '../../Constants/Constants';

export default class ComposeMessageForm extends Component {
  state = {
    message: '',
  }

  handleChange = (message) => {
    this.setState({ message });
  }

  sendMessage = async () => {
    const { message } = this.state;
    if(!message) return;
    const { from, to, updateConversation } = this.props;
    const sent = new Date();
    const sentSeconds = Math.floor(sent.getTime() / 1000);
    const newMessage = {
      from,
      contents: message,
      sent,
    };
    const sentMessage = {
      to,
      contents: message,
      sent: { seconds: sentSeconds },
    };
    try {
      await firebase.firestore().collection('users').doc(to).collection('inbox').add(newMessage);
      this.setState({message: '' });
      updateConversation(to, { contents: message, timestamp: { seconds: sentSeconds }, sender: true });
      await this.saveSentMessage(sentMessage);
    } catch(error) { console.log(error); }
  }

  saveSentMessage = async (newMessage) => {
    const { from } = this.props;
    try{
      const stringySavedMessages = await AsyncStorage.getItem(from);
      const savedMessages = JSON.parse(stringySavedMessages);
      savedMessages.sent.push(newMessage);
      await AsyncStorage.setItem(from, JSON.stringify(savedMessages));
    } catch(err) {console.error({ err })}   
  }
  
  render() {
    return (
      <View style={styles.container}>
        <TextInput style={styles.input} onChangeText={this.handleChange} value={this.state.message} placeholder="Send a message..."/>
        <TouchableOpacity style={styles.sendBtnContainer} onPress={this.sendMessage}>
          <Image source={sendImg} style={{width: 50, height: 50}}/>
        </TouchableOpacity>
      </View>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    backgroundColor: 'white',
    width: Dimensions.get('window').width * .9,
    display: 'flex',
    flexDirection: 'row',
    position: 'absolute',
    height: 60,
    top: Dimensions.get('window').height - 90,
    zIndex: 5,
    borderRadius: 6,
  },
  input: {
    width: '80%',
    textAlign: 'center',
    fontSize: 18,
  },
  sendBtnContainer: {
    backgroundColor: constants.primaryBgColor,
    borderColor: 'white',
    borderWidth: 2,
    borderRadius: 6,
    width: '20%',
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
  }
});

