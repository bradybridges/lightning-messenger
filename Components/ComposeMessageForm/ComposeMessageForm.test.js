import React from 'react';
import { shallow } from 'enzyme';
import ComposeMessageForm from './ComposeMessageForm';

describe('ComposeMessageForm', () => {
  let wrapper;
  const mockUpdateConversation = jest.fn();

  beforeEach(() => {
    wrapper = shallow(
      <ComposeMessageForm 
        to="receiver" 
        from="sender" 
        mockUpdateConversation={mockUpdateConversation}
      />
    );
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should have state containing a property of message', () => {
    expect(wrapper.state('message')).toEqual('');
  });

  it('handleChange should update with a given message', () => {
    const mockMessage = 'Hello there';
    wrapper.instance().handleChange(mockMessage);
    expect(wrapper.state('message')).toEqual(mockMessage);
  });

  it('sendMessage should do nothing if there is no message to send', () => {
    wrapper.instance().sendMessage();
    expect(mockUpdateConversation).not.toHaveBeenCalled();
  });
});