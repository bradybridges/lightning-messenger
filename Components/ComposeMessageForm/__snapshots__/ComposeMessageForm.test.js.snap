// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComposeMessageForm should match snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "borderRadius": 6,
      "display": "flex",
      "flexDirection": "row",
      "height": 60,
      "position": "absolute",
      "top": 1244,
      "width": 675,
      "zIndex": 5,
    }
  }
>
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    placeholder="Send a message..."
    rejectResponderTermination={true}
    style={
      Object {
        "fontSize": 18,
        "textAlign": "center",
        "width": "80%",
      }
    }
    underlineColorAndroid="transparent"
    value=""
  />
  <TouchableOpacity
    activeOpacity={0.2}
    onPress={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#000000d6",
        "borderColor": "white",
        "borderRadius": 6,
        "borderWidth": 2,
        "display": "flex",
        "justifyContent": "center",
        "width": "20%",
      }
    }
  >
    <Image
      source={1}
      style={
        Object {
          "height": 50,
          "width": 50,
        }
      }
    />
  </TouchableOpacity>
</View>
`;
